skxray/cdi.py

```diff
@@ -119,9 +125,339 @@ def convolution(array1, array2):
+            ny/2-sup_radius: ny/2+sup_radius] = 1
+    elif len(shape_v) == 3:
+        nx, ny, nz = shape_v
+        sup[nx/2-sup_radius: nx/2+sup_radius,
+            ny/2-sup_radius: ny/2+sup_radius,
+            nz/2-sup_radius: nz/2+sup_radius] = 1
+    return sup
+
+
+def generate_disk_support(sup_radius, shape_v):
+    """
+    Generate support area as a disk for either 2D or 3D cases.
+
+    Parameters
+    ----------
+    sup_radius : float
+        radius of support
+    shape_v : list
+        shape of diffraction pattern, which can be either 2D or 3D case.
+
+    Returns
+    -------
+    sup : array
+        support with a disk area
+    """
+    sup = np.zeros(shape_v)
+    dummy = _dist(shape_v)
+    sup_index = np.where(dummy <= sup_radius)
+    sup[sup_index] = 1
+    return sup
+
+
+def cdi_recon(diff_array, sample_obj, sup,
+              beta=1.15, start_ave=0.8, pi_modulus_flag='Complex',
+              sw_flag=True, sw_sigma=0.5, sw_threshold=0.1, sw_start=0.2,
+              sw_end=0.8, sw_step=10, n_iterations=1000):
+    """
+    Run reconstruction with difference map algorithm.
+
+    Parameters
+    ---------
+    diff_array : array
+        diffraction pattern from experiments
+    sample_obj : array
+        initial sample with phase
+    sup : array
+        initial support
+    beta : float, optional
+        feedback parameter for difference map algorithm.
+        default is 1.15.
+    start_ave : float, optional
+        define the point to start doing average.
+        default is 0.8.
+    pi_modulus_flag : str, optional
+        'Complex' or 'Real', defining the way to perform pi_modulus calculation.
+        default is 'Complex'.
+    sw_flag : Bool, optional
+        flag to use shrinkwrap algorithm or not.
+        default is True.
+    sw_sigma : float, optional
+        gaussian width used in sw algorithm.
+        default is 0.5.
+    sw_threshold : float, optional
+        shreshold cut in sw algorithm.
+        default is 0.1.
+    sw_start : float, optional
+        at which point to start to do shrinkwrap.
+        defualt is 0.2
+    sw_end : float, optional
+        at which point to stop shrinkwrap.
+        defualt is 0.8
+    sw_step : float, optional
+        the frequency to perform sw algorithm.
+        defualt is 10
+    n_iterations : int, optional
+        number of iterations to run.
+        default is 1000.
+
+    Returns
+    -------
+    obj_ave : array
+        reconstructed sample object
+    error_dict : dict
+        Error information for all iterations. The dict keys include
+        obj_error, diff_error and sup_error. Obj_error is a list of
+        the relative error of sample object. Diff_error is calculated as
+        the difference between new diffraction pattern and the original
+        diffraction pattern. And sup_error stores the size of the
+        sample support.
+    """
+
+    diff_array = np.array(diff_array)     # diffraction data
```

Comment:
I am guessing that there will be a major performance gain due to reduced memory allocation.  The reason that you need the two fftshifts in the fft/ifft_helpers is because the diffraction image is [-w, 0, w] where as the order that fft yieds/expects is [0, 1, ... w, -w, -(w-1), -1].  If you re-order the diffraction image on the way in all of your internal functions can assume the 'hardware' ordering.

ex

```python
In [30]:  n = np.random.rand(200, 200)

In [32]: %timeit np.fft.fft(n)
1000 loops, best of 3: 493 µs per loop

In [33]: %timeit np.fft.fftshift(np.fft.fft(np.fft.fftshift(n)))
1000 loops, best of 3: 1.11 ms per loop

In [34]: n = np.random.rand(2000, 2000)

In [35]: %timeit np.fft.fft(n)
10 loops, best of 3: 72.1 ms per loop

In [36]: %timeit np.fft.fftshift(np.fft.fft(np.fft.fftshift(n)))
10 loops, best of 3: 109 ms per loop

```